kyo-core/shared/src/main/scala/kyo/envs.scala

```diff
@@ -26,35 +38,40 @@ object Envs:
         def apply[T: Flat, S, VS, VR](env: V)(value: T < (Envs[VS] & S))(
             using
             HasEnvs[V, VS] { type Remainder = VR },
-            Tag[Envs[V]]
+            Tag[V]
         ): T < (S & VR) =
-            envs[V].handle(handler[V])(env, value).asInstanceOf[T < (S & VR)]
+            given Tag[Envs[V]] = makeEnvsTag[V]
+            envs[V].handle(handler[V])(TypeMap(env), value).asInstanceOf[T < (S & VR)]
+        end apply
     end RunDsl
 
     def run[V >: Nothing]: RunDsl[V] =
         new RunDsl[V]
 
-    private def handler[V]: ResultHandler[V, Const[Unit], Envs[V], Id, Any] =
-        cachedHandler.asInstanceOf[ResultHandler[V, Const[Unit], Envs[V], Id, Any]]
+    def provide[V >: Nothing, T: Flat, S, VS, VR](env: TypeMap[V])(value: T < (Envs[VS] & S))(
+        using
+        HasEnvs[V, VS] { type Remainder = VR },
+        Tag.Intersection[V]
+    ): T < (S & VR) =
+        given Tag[Envs[V]] = makeEnvsTag[V]
+        envs[V].handle(handler[V])(env, value).asInstanceOf[T < (S & VR)]
+    end provide
+
+    private def handler[V]: ResultHandler[TypeMap[V], Tag, Envs[V], Id, Any] =
+        cachedHandler.asInstanceOf[ResultHandler[TypeMap[V], Tag, Envs[V], Id, Any]]
 
     private val cachedHandler =
-        new ResultHandler[Any, Const[Unit], Envs[Any], Id, Any]:
-            def done[T](st: Any, v: T)(using Tag[Envs[Any]]) = v
-            def resume[T, U: Flat, S2](st: Any, command: Unit, k: T => U < (Envs[Any] & S2))(using Tag[Envs[Any]]) =
-                Resume(st, k(st.asInstanceOf[T]))
-
-    /** An effect `Envs[VS]` includes a dependency on `V`, and once `V` has been handled, `Envs[VS]` should be replaced by `Out`
-      *
-      * @tparam V
-      *   the dependency included in `VS`
-      * @tparam VS
-      *   all of the `Envs` dependencies represented by type intersection
-      */
-    sealed trait HasEnvs[V, VS]:
-        /** Remaining effect type, once the `V` dependency has been provided
-          */
+        new ResultHandler[TypeMap[Any], Tag, Envs[Any], Id, Any]:
+            override def accepts[T](st: TypeMap[Any], command: Tag[T]): Boolean =
+                st.tag <:< command
```

Comment:
Perhaps we should make tag a lazy val? Or I can make the state a tuple of the type map and its intersection tag.

Reply:
Agree! Avoiding the allocations to obtain the tag is important here since it's a hot path. I think a lazy val would work well since the tuple allocation should be more expensive.